### PR TITLE
add ScrollViewContentObserver

### PR DESCRIPTION
**Pull request by noahemmet**
### Checklist
- [x] All new and existing tests pass
- [x] I've read the [Contribution Guidelines](https://github.com/google/science-journal-ios/blob/master/CONTRIBUTING.md)
- [x] I've read [Change Limitations](https://github.com/google/science-journal-ios/blob/master/CHANGE_LIMITATIONS.md)

### Motivation and Context

The `ActionArea.BarViewController` wasn't updating its appearance (raising or lowering a drop shadow) after certain events, like pushing/popping view controllers, or adding/removing/editing entries. (Only scrolling its content would update correctly.)

### Description

This PR:
- Adds an observer for `scrollView.contentSize` in addition to the `.contentOffset` when determining the `ActionArea`'s appearance.
- Moves the observer logic out of the scrollView delegate callbacks and into a `ScrollViewContentObserver` object.

I've tested various pushing/popping/editing of the `TrialDetailViewController` contents; if you think of any other edge cases, I'd be happy to test those as well.

### CODE REVIEW

File: ScienceJournal/UI/MaterialHeaderViewController.swift [hunk 1]
```diff
@@ -28,7 +28,8 @@ class MaterialHeaderViewController: ScienceJournalViewController, UIScrollViewDe
 
   let appBar = MDCAppBar()
   var trackedScrollView: UIScrollView? { return nil }
-  @objc dynamic private(set) var isContentOutsideOfSafeArea: Bool = false
+  @objc private(set) lazy var scrollViewContentObserver
+    = ScrollViewContentObserver(scrollView: trackedScrollView)
```
Comment (garrettmurray) [inline ScienceJournal/UI/MaterialHeaderViewController.swift]: Nit: Formatting-wise in SJ we do the `=` on the preceding line instead, like:

```
@objc private(set) lazy var scrollViewContentObserver =
  ScrollViewContentObserver(scrollView: trackedScrollView)
```

File: ScienceJournal/UI/ObserveViewController.swift [hunk 2]
```diff
@@ -109,7 +109,8 @@ open class ObserveViewController: ScienceJournalCollectionViewController, ChartC
   let recordingManager: RecordingManager
   var recordingTrial: Trial?
   let sensorController: SensorController
-  @objc dynamic private(set) var isContentOutsideOfSafeArea: Bool = false
+  @objc private(set) lazy var scrollViewContentObserver
+    = ScrollViewContentObserver(scrollView: collectionView)
```
Comment (garrettmurray) [inline ScienceJournal/UI/ObserveViewController.swift]: Same comment here.

File: ScienceJournal/UI/MaterialHeaderViewController.swift (hunk 1 above)
Comment (noahemmet) [reply ScienceJournal/UI/MaterialHeaderViewController.swift]: Ah, thanks; done.
